core/tracing/CHANGELOG.md

```diff
@@ -4,6 +4,53 @@ All notable changes to the tracing interface will be documented in this file.
 
 ## [Unreleased]
 
+The tracing interface has been extended with backwards-compatible changes to support more use-cases and simplify tracer code. The most notable changes are state read hooks as well a state journaling library which emits events when a call is reverted.
+
+### New methods
+
+- `OnReorg(reverted []*types.Block)`: This hook is called when a reorg is detected. The `reverted` slice contains the blocks that are no longer part of the canonical chain.
```

Comment:
Here types block is very very heavy. You should *at most* pass headers and allow chain access to pull the blocks on demand (chain access in someconstructor, ha)

Reply:
On second thought what is the issue? it is a slice so passed by reference and the memory can be freed as soon as `OnReorg` processing is done.

Reply:
Ugh, this is annoying. So reorg in the blockchain at some point in the past used to collect blocks. Turned out that sometimes it became insanely heavy and we've switched so it operates on headers. I guess later someone refactored it back to operate on blocks again. This is an issue when you do setHead or any similar operation; of even if finality fails for a while and you have blocks reorging back and forth. It's very very bad to pull all the block in from disk IMO.

CC @holiman @rjl493456442 ?

Reply:
I agree. I don't particularly recall switching from headers to blocks....